docs/secrets.mdx

```diff
@@ -0,0 +1,630 @@
+### Secrets, Versions and Values
+
+Secrets are a named resource, identified by their name. Secrets are also versioned, so when you access a
+secret's value you are accessing a particular version of that secret.
+
+This schema is illustrated below for a secret `db.password` with two versions:
+
+```
++- Secret [ 'db.password' ]
+   |
+   +- SecretVersion [ '7F5F86D0-D97F-487F-A5A0-11BAAD00F777' ]
+   |  |
+   |  +- SecretValue [ 'bleak_dearest_hanged_reigns' ]
+   |
+   +- SecretVersion [ '0581BBD9-C67F-4E8F-849D-38E52CAEE0EB' ]
+      |
+      +- SecretValue [ 'crummy_goofed_caddy_radiant' ]
+```
+
+By using versioned secrets you can support password or key rotation policies.
+
+### References and Secret Access
+
+The Secret and SecretVersion objects are light weight references.
+
+To access the actual secret value (data) you use the SecretVersion `access()` method which will make a server
+call returning a SecretValue object.
+
+To get the latest version of a secret you use the `latest()` method on the Secret object.
+
+## Usage
+
+Using one of the provided Nitric SDKs for your language is the easiest way to work with
+Secrets.
+
+### Put
+
+To create a new SecretValue use the Secret `put` method.
+
+<CodeExamples 
+	languages={[
+	{
+		label: "Node",
+		value: "node"
+	},
+	{
+		label: "Python",
+		value: "python"
+	},
+	{
+		label: "Java",
+		value: "java"
+	},
+	{
+		label: "Go",
+		value: "go"
+	},
+	{
+		label: "PHP",
+		value: "php"
+	},
+	{
+		label: "C#",
+		value: "c-sharp"
+	}
+	]}
+	defaultLang="node"
+>
+<CodeExample lang="node">
+
+```typescript
+import { storage } from '@nitric/sdk';
+
+const secrets = new Secrets();
+
+const newPassword = 'qxGJp9rWMbYvPEsNFXzukQa!';
+
+// Store the new password value, making it the latest version
+secrets.secret('database.password').put(newPassword);
+```
+
+</CodeExample>
+<CodeExample lang="python">
+
+```
+Secrets support is under development...
+```
+
+</CodeExample>
+<CodeExample lang="java">
+
+```java
+import io.nitric.api.secret.Secrets;
+
+var newPassword = "qxGJp9rWMbYvPEsNFXzukQa!";
+
+// Store the new password value, making it the latest version
+Secrets.secret("database.password").putAsText(newPassword);
+```
+
+</CodeExample>
+<CodeExample lang="go">
+
+```go
+import (
+    "github.com/nitrictech/go-sdk/api/secrets"
+)
+
+ss, err := secrets.New()
+if err != nil {
+	// handle error
+}
+
+newPassword := "qxGJp9rWMbYvPEsNFXzukQa!";
+
+// Store the new password value, making it the latest version
+_, err := ss.Secret("database.password").Put([]byte(newPassword))
+
+if err != nil {
+	// Handle error
+}
+```
+
+</CodeExample>
+<CodeExample lang="php">
+
+```
+Secrets support is under development...
+```
+
+</CodeExample>
+
+<CodeExample lang="c-sharp">
+
+```csharp
+using Nitric.Api.Secret;
+
+var secrets = new Secrets();
+
+var newPassword = "qxGJp9rWMbYvPEsNFXzukQa!";
+
+// Store the new password value, making it the latest version
+secrets.Secret("database.password").Put(newPassword);
```

Comment:
Note that `.` is not currently in the regex for secret name validation, super easy to add it, but will need to be there for these examples to be valid :).

Reply:
The current GRCPC validation allows `.`  from what I can see. There is a test case for this.

Reply:
I put to that in after I wrote the initial comment 🙂

Reply:
Sweet, thanks